src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,217 @@
+
+#include "ray/core_worker/transport/direct_actor_transport.h"
+#include "ray/common/task/task.h"
+
+using ray::rpc::ActorTableData;
+
+namespace ray {
+
+bool HasByReferenceArgs(const TaskSpecification &spec) {
+  for (int i = 0; i < spec.NumArgs(); ++i) {
+    if (spec.ArgIdCount(i) > 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+CoreWorkerDirectActorTaskSubmitter::CoreWorkerDirectActorTaskSubmitter(
+    boost::asio::io_service &io_service, gcs::RedisGcsClient &gcs_client,
+    CoreWorkerObjectInterface &object_interface)
+    : io_service_(io_service),
+      gcs_client_(gcs_client),
+      client_call_manager_(io_service),
+      store_provider_(
+          object_interface.CreateStoreProvider(StoreProviderType::LOCAL_PLASMA)) {
+  RAY_CHECK_OK(SubscribeActorUpdates());
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(
+    const TaskSpecification &task_spec) {
+  if (HasByReferenceArgs(task_spec)) {
+    return Status::Invalid("direct actor call only supports by-value arguments");
+  }
+
+  RAY_CHECK(task_spec.IsActorTask());
+  const auto &actor_id = task_spec.ActorId();
+
+  const auto task_id = task_spec.TaskId();
+  const auto num_returns = task_spec.NumReturns();
+
+  auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
+  request->set_task_id(task_spec.TaskId().Binary());
+  request->mutable_task_spec()->CopyFrom(task_spec.GetMessage());
+
+  std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+  auto iter = actor_states_.find(actor_id);
+  if (iter == actor_states_.end() || iter->second == ActorTableData::RECONSTRUCTING) {
```

Comment:
Is it possible the actor never shows up in actor_states_ since it was a bad handle or something?

Reply:
Sorry I don't quite get how it relates to actor handle, may you please give an example?  The `actor_states_` is updated by GCS notification, as long as the actor is created successfully, it should appear in `actor_states_`.

Reply:
I'm thinking of something like unpickling a random invalid actor handle -- would be nice to have a timeout that if the actor state doesn't show up we mark it as invalid. But this is minor.

Reply:
I see. Will add a TODO comment, think this probably can be addressed later.